relay-metrics/src/aggregation.rs

```diff
@@ -711,6 +759,78 @@ impl Bucket {
     }
+
+    /// Splits this bucket if its estimated serialization size exceeds a threshold.
+    ///
+    /// There are three possible return values:
+    ///  - `(Some, None)` if the bucket fits entirely into the size budget. There is no split.
+    ///  - `(None, Some)` if the size budget cannot even hold the bucket name and tags. There is no
+    ///    split, the entire bucket is moved.
+    ///  - `(Some, Some)` if the bucket fits partially. Remaining values are moved into a new bucket
+    ///    with all other information cloned.
+    ///
+    /// This is an approximate function. The bucket is not actually serialized, but rather its
+    /// footprint is estimated through the number of data points contained. See
+    /// [`estimated_size`](Self::estimated_size) for more information.
+    fn split_at(mut self, size: usize) -> (Option<Bucket>, Option<Bucket>) {
+        // If there's enough space for the entire bucket, do not perform a split.
+        if size >= self.estimated_size() {
+            return (Some(self), None);
+        }
+
+        // If the bucket key can't even fit into the remaining length, move the entire bucket into
+        // the right-hand side.
+        let own_size = self.estimated_own_size();
+        if size < own_size {
+            return (None, Some(self));
+        }
+
+        // Perform a split with the remaining space after adding the key. We assume an average
+        // length of 8 bytes per value and compute the number of items fitting into the left side.
+        let split_at = (size - own_size) / AVG_VALUE_SIZE;
+
+        match self.value {
+            BucketValue::Counter(_) => (None, Some(self)),
+            BucketValue::Distribution(ref mut distribution) => {
+                let org = std::mem::take(distribution);
+                let mut new_bucket = self.clone();
+
+                let mut iter = org.iter_values();
+                self.value = BucketValue::Distribution((&mut iter).take(split_at).collect());
+                new_bucket.value = BucketValue::Distribution(iter.collect());
+
+                (Some(self), Some(new_bucket))
+            }
+            BucketValue::Set(ref mut set) => {
+                let org = std::mem::take(set);
+                let mut new_bucket = self.clone();
+
+                let mut iter = org.into_iter();
+                self.value = BucketValue::Set((&mut iter).take(split_at).collect());
+                new_bucket.value = BucketValue::Set(iter.collect());
+
+                (Some(self), Some(new_bucket))
+            }
+            BucketValue::Gauge(_) => (None, Some(self)),
+        }
+    }
+
+    /// Estimates the number of bytes needed to serialize the bucket without value.
+    ///
+    /// Note that this does not match the exact size of the serialized payload. Instead, the size is
+    /// approximated through tags and a static overhead.
+    fn estimated_own_size(&self) -> usize {
```

Comment:
nit: I find `own` here rather confusing.  maybe `estimate_base_size` could be better?